R/getTimeDivergence.R

```diff
@@ -0,0 +1,98 @@
+#' steps. It can be 1 or higher than 1.
+#' @param transposed logical scalar assigning samples to rows if they are in
+#' columns. If samples are not in rows, the default value is applied.
+#' (default: \code{transposed = FALSE})
+#'
+#' @return a matrix containing the beta diversity and time difference
+#' between samples in n time steps
+#'
+#' @importFrom mia calculateDistance
+#'
+#' @examples
+#' library(microbiomeDataSets)
+#' tse <- GrieneisenTSData()
+#'
+#' BaboonDivergence <- getTimeDivergence(tse, sample_field = "baboon_id",
+#'                                     sample_id = "Baboon_1" ,
+#'                                     time_field = "collection_date",
+#'                                     time_interval = 1,
+#'                                     transposed = FALSE )
+#'
+#'@export
+getTimeDivergence <- function(se, sample_field, sample_id, time_field, time_interval, transposed = FALSE ){
+
+    df <-as.data.frame(colData(se)[, c(sample_field, time_field)])
+
+    #per sample
+    sample_df <- df[which((df[, sample_field] == sample_id) == TRUE),]
+
+    mat <- data.matrix(sample_df)
+
+    #time in increasing order
+    mat[, time_field] <- mat[, time_field][order(mat[, time_field], decreasing = FALSE)]
+
+     tt <- sapply(seq_len(nrow(mat)), FUN = function(i){
+      k <- c(i, i + time_interval)
+      i <- i + 1
+      k
+     })
+
+    #total number of sample pairs
+    total <- nrow(mat) - time_interval
+
+    location <- tt[,seq_len(total)]
+
+    #location matrix turned into list object
+    if(length(location) > 2){
+        list <- lapply(seq_len(ncol(location)), function(i) location[,i])
+
+        samplename <-lapply(seq_len(length(list)), function(i){ mat[, sample_field][list[[i]]]})
+
+        time <-lapply(seq_len(length(list)), function(i){ mat[, time_field][list[[i]]]})
+
+        #related samples are chosen from assay
+        list_t <- lapply(seq_len(length(time)), function(i){ as.matrix(assay(se)[,names(samplename[[i]])])})
+
+        if(!transposed){
+          list_t <- lapply(seq_len(length(list_t)), function(i){
+            t(list_t[[i]])
+          })
+        }
+
+        timedivergence_n <- vapply(seq_len(length(list_t)),
+                                   FUN = function(i) {calculateDistance(list_t[[i]])},
+                                   FUN.VALUE = 1)
+
+
+        timedifference_n <- vapply(seq_len(length(timedivergence_n)),
+                                   FUN = function(i) {diff(time[[i]])},
+                                   FUN.VALUE = 1)
+    } else {
+        samplename <- mat[, sample_field][location]
+        time <- mat[, time_field][location]
+        mat_t <- as.matrix(assay(se)[,names(samplename)])
+        if(!transposed){
+          mat_t <- t(mat_t)
+        }
+        timedivergence_n <- calculateDistance(mat_t)
+        timedifference_n <- diff(time)
+    }
+
+    return(cbind(timedivergence_n,timedifference_n))
```

Comment:
It would be handy to have additionally a function `estimateTimeDivergence` that adds the output from `getTimeDivergence` to input `SE` object's `colData` directly. In the same way than in `mia::estimateDiversity`.